### PR TITLE
POC: Implement simple front-end pagination for the Notebook

### PR DESCRIPTION
**Pull request by lyzadanger**
Part of https://github.com/hypothesis/client/issues/2924

This PR is a POC for front-end pagination of threads in the Notebook view, to make large sets of threads usable. Right now, larger sets of threads cause uncomfortable amounts of scrolling, and it’s hard to keep track of one’s general place.

Code is WIP quality.

I am looking for high-level implementation feedback and any usability red flags here. This behavior was demo’ed to the dev team on February 8, 2021.

### General Premises and Notes

* Pagination is based on _top-level threads_, not annotations. That is, it controls the number of annotation thread cards that will show up on a single page of results. While this will generally limit the (vertical) length of pages, it is technically not guaranteed that pages will stay under a certain vertical height, as it depends on the content of annotations and the depth of reply nesting. _I believe this to be okay_.
* Pagination page size is dialed way down in this POC: 10 root threads per page. This is too small, but is meant to make playing around with this easier.
* This POC aims to solve the very basic first concerns of simple pagination. In scope are: how front-end threads are paginated (logically) and how the pagination controls look and feel to the user. There’s a lot more to do here—effectively everything else is out of scope (URL hashing/navigation; paginating back-end result sets).

### Pagination UX

Pagination UX is derived from review of UX articles, UX example sites (e.g. dribbble.com), inspiration from sites that I feel do pagination well (Etsy, Nordstrom), and checking in on the heavy-hitters (Google).

Exploratory sketches are on [Balsamiq](https://balsamiq.cloud/s9ipy1g/pb2ycd2/r3DA1) (credentials in shared password manager for H team members).

I decided on trying out page-based navigation (versus infinite loading) for these reasons:

* It’s more intuitive (IMHO) with a bounded set of results. Infinite loading feels more naturally suited for a changing, “stream” of results.
* Page-based pagination constrains vertical size and scrolling, such that top-of-view and filters stay reasonably well at hand.
* It’s easier (or, even, possible) to jump forward in or backward results without having to load everything between or scroll through a whole lot.
* Page-based pagination can be more naturally mapped to URLs, and it’s easier to support the back button (NB: URL/navigation support has not yet been implemented).

For the main navigation controls, the goals are:

* Provide relative controls (prev/next) where relevant. “Prev” does seem to be a widely-used text string for buttons (versus “Previous”) and it makes the buttons considerably more visually balanced.
* Always show numbered links/buttons to first and last page of results.
* Always show numbered link/button for the active page.
* Show numbered links/buttons for the pages “near” the currently-active page, up to five numbered links/buttons total.

This is more easily shown by examples:

image

image

image

image

image

Smaller page set:

image

### Implementation

There are three main logical concerns for this pagination:

1. We need to manage state about pagination: what page we’re on right now, responding to changes to that page, and resetting back to page 1 on certain other state changes. (`NotebookView`, but could be moved to the store if we wanted).
2. We need to map pagination onto the total set of results (threads), determine the total number of pages and show the right threads for the current page. (`PaginatedThreadList`)
3. We need to present pagination controls and information to the user. (`PaginationNavigation`).

#### In more depth:

* `NotebookView`: Manages state related to which pagination page is currently active (provides page-changing callback to child components). Responds to page changes.
    
    This logic is in this component because of the proximity to the loading of annotations, which affects pagination. If we moved pagination-state management to the store, we’d have more independence about where this logic lives.

    `NotebookView`’ would need to be updated if we changed how pagination state is managed, or if we changed what things cause a pagination “reset.”

* `PaginatedThreadList`: Determines which `threads` are meaningful for pagination calculation (currently: all _visible_ top-level threads). Determines how many pages of results are in the current set and which threads should be visible on the current page.

     Passes currently-visible threads onto `ThreadList` and pagination metadata to `PaginationNavigation` .

     `PaginatedThreadList` would need to be updated if the relationship between root-level `threads` and pagination changed (e.g. if `threads` were not the total set of results). It’s expected that this would happen when we start integrating back-end support for filtering and thread retrieval.

* `PaginationNavigation`: Takes `currentPage`, `totalPages` and the `onChangePage` callback and renders pagination controls.

     `PaginationNavigation` would need to be updated if we changed our minds about how we wanted pagination-control behavior or appearance to change, but is independent of other pagination logic.

* `NotebookResultCount`: Changed to be able to render what page of results is showing. Also refactored lightly to remove dependency on `storeProxy` (this is consistent with other child components of `NotebookView`.

Changes since informal demo to team:

* Pagination page number will be rendered at the top of the view next to total results count when the page number is > 1. This is directly cribbed from Google (try it!) and I like it because:
	* It doesn’t clutter the top of the view with extra information (“Page 1”) when it’s not relevant.
	* It obviates the need for the results-count component to know about the total number of pages of results (which is calculated in a sibling component).

image

image

In the not-too-distant future, on wider screens, we'll be moving the filters to the left of results, and we may wish de-clutter this a little bit.

### Supporting changes

* `ThreadList` takes an Array of `Thread` objects instead of a single "root" Thread.
* `Button` takes an icon-position prop

### Known further steps

- [ ] Likely: Extract page-button-number logic from `PaginationNavigation` into a utility and comment it more clearly
- [ ] Address spacing issues with the `arrow-right` icon (see `PaginationNavigation.scss` for current workaround)
- [ ] Write tests (obviously)

### CODE REVIEW

File: src/sidebar/components/Button.js [hunk 1]
```diff
@@ -108,6 +112,7 @@ Button.propTypes = {
   buttonText: propTypes.string,
   className: propTypes.string,
   icon: requiredStringIfButtonTextMissing,
+  iconPosition: propTypes.string,
```
Comment (lyzadanger) [inline src/sidebar/components/Button.js:115]: propTypes can't die fast enough.

File: src/sidebar/components/NotebookResultCount.js [hunk 2]
```diff
@@ -1,9 +1,19 @@
+import propTypes from 'prop-types';
+
 import useRootThread from './hooks/use-root-thread';
-import { useStoreProxy } from '../store/use-store';
 import { countVisible } from '../helpers/thread';
 
 import Spinner from './Spinner';
 
+/**
+ * @typedef NotebookResultCountProps
```
Comment (lyzadanger) [inline src/sidebar/components/NotebookResultCount.js:9]: `NotebookResultCount` takes several more props as of these changes, most of which are unrelated to the changes at hand but an attempt to make component "sophistication" balanced. In this case, the removal of child component dependencies on `storeProxy`.

File: src/sidebar/components/NotebookResultCount.js [hunk 3]
```diff
@@ -35,21 +46,22 @@ function NotebookResultCount() {
       {isLoading && <Spinner />}
       {!isLoading && (
         <h2>
-          {!hasResults && <span>No results</span>}
-          {hasResults && hasAppliedFilter && (
-            <span>
+          {currentPage > 1 && <span>Page {currentPage} of </span>}
```
Comment (lyzadanger) [inline src/sidebar/components/NotebookResultCount.js:49]: This is the only meaningful change in this component.

File: src/sidebar/components/NotebookView.js [hunk 4]
```diff
@@ -21,37 +23,88 @@ import ThreadList from './ThreadList';
  */
 function NotebookView({ loadAnnotationsService }) {
   const store = useStoreProxy();
+
+  const filters = store.getFilterValues();
```
Comment (lyzadanger) [inline src/sidebar/components/NotebookView.js:27]: `NotebookView` has been made into a "clearing house" of store-state-related management, versus having this distributed throughout its children.

File: src/sidebar/components/NotebookView.js [hunk 5]
```diff
@@ -21,37 +23,88 @@ import ThreadList from './ThreadList';
...
+    // within the Notebook to change the `focusedGroup`. If the focused group
+    // is changed within the sidebar and the Notebook re-opened, an entirely
+    // new iFrame/app is created. This will need to be revisited.
     store.setSortKey('Newest');
     if (focusedGroup) {
       loadAnnotationsService.load({
         groupId: focusedGroup.id,
         maxResults: 5000,
       });
+      setPaginationPage(1);
```
Comment (lyzadanger) [inline src/sidebar/components/NotebookView.js:57]: The idea is that we'd want to reset pagination after reloading new results. In real life right now, this block only executes once at app startup (see notes in comments above), but in the future, that will likely change.

File: src/sidebar/components/PaginationNavigation.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+ * Determine the set of page number buttons to show in the navigation
+ * controls, in addition to "next" and "back" buttons.
+ *
+ * @param {number} totalPages
+ * @param {number} currentPage
+ * @return {PageNumber[]} Set of navigation page options to show. `null`
+ *   values represent gaps in the sequence of pages, to be represented later
+ *   as ellipses (...)
+ */
+function availablePageButtons(currentPage, totalPages) {
```
Comment (lyzadanger) [inline src/sidebar/components/PaginationNavigation.js:17]: As noted in PR description, I'll likely pull this into a util

File: src/sidebar/components/PaginationNavigation.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            title="Go to previous page"
+            onClick={e => changePageTo(currentPage - 1, e.target)}
+          />
+        )}
+      </div>
+      <ul className="PaginationNavigation__pages">
+        {pageNumbers.map((page, idx) => (
+          <li key={`page-${idx}`}>
+            {page === null ? (
+              <div className="PaginationNavigation__gap">...</div>
```
Comment (lyzadanger) [inline src/sidebar/components/PaginationNavigation.js:100]: We should make sure this is OK, vis-a-vis a11y.

File: src/styles/sidebar/components/PaginationNavigation.scss [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+@use "../../variables" as var;
+
+.PaginationNavigation {
+  @include utils.font--large;
+  @include layout.row;
+
+  &__relative {
+    // Set a constant width for the previous and next buttons so that
+    // there is no jumping around as they appear and disappear.
+    width: 5em;
```
Comment (lyzadanger) [inline src/styles/sidebar/components/PaginationNavigation.scss:13]: I feel like there's a better way to do this...

**Comment by robertknight** [comment]
What follows are some brief notes from an initial hour or so of playing around with this using the browser extension (so I can test with more realistic data) and perusing the code.

Overall I think this looks like a good first stab at pagination. The rationale for using pagination vs infinite loading makes sense. The UI controls for navigating between pages look good. I don't think the result count labeling is quite right yet. More comments on that below. The implementation structure looks pretty sensible and straightforward from an initial read-through. I agree with the choices to make the pagination control re-usable and to hoist some of state upwards so that the child components don't need to know about it.

Comments on a few specifics:

- The pagination control looks good. One possible usability query I have is whether we should optimize for quickly flipping between several pages. Right now the UI is not optimized for this because of the scroll-to-top behavior when switching pages and because the Prev/Next actions are not next to each other. To clarify, I'm not suggesting a change here, just raising something that I think we need to look for user feedback on.
- The hit area of each page number button could be increased. There is currently dead space between each one.
- The "Page $X of $Y threads" text in the top right looks wrong to me, because $X and $Y are different units. Showing "Threads $START to $END of $TOTAL" might work better? We could perhaps remove the total annotation count to make room for that. See Gmail's mail counter for a reference.
- When navigating through a large number of results, I definitely think it would be useful to have controls at the top, for the use case where someone is trying to skip to annotations made in some rough time period for example. These don't necessarily need to be the full controls, just Next/Prev arrows would be useful.

Some general comments unrelated to this PR, but which occurred to me while testing it:

I think there may be some confusion caused by the fact that the navigation controls deal with threads, but the filter controls deal with annotations and replies. For example, filtering by user X and then clicking "Show N more in conversation" to see the rest of the discussion results in the view showing "Page P of N threads, and Z more" where "P" is a page number, "N" is a thread count and "Z" is a reply count.

A related issue is that I think using the term "annotations" to include "replies" in this interface may be confusing. We represent replies as a type of annotation internally, but that may not (probably isn't?) how the user thinks of them. 

Lastly perhaps there should be an option to toggle whether replies are included in the results, or only top-level annotations/notes.

**Comment by lyzadanger** [comment]
> I think there may be some confusion caused by the fact that the navigation controls deal with threads, but the filter controls deal with annotations and replies. For example, filtering by user X and then clicking "Show N more in conversation" to see the rest of the discussion results in the view showing "Page P of N threads, and Z more" where "P" is a page number, "N" is a thread count and "Z" is a reply count.

> A related issue is that I think using the term "annotations" to include "replies" in this interface may be confusing. We represent replies as a type of annotation internally, but that may not (probably isn't?) how the user thinks of them.

I've been trying to get attention to both of these issues for months. I don't really know what to tell yah.

**Comment by lyzadanger** [comment]
> For example, filtering by user X and then clicking "Show N more in conversation" to see the rest of the discussion results in the view showing "Page P of N threads, and Z more" where "P" is a page number, "N" is a thread count and "Z" is a reply count.

I don't think this combination is possible? When filtering by anything, the result count goes from threads to "results". So you might see "Page P of N results (and Z more)". But you wouldn't see the term "threads" in this context.

**Comment by lyzadanger** [comment]
To address your points!

> The pagination control looks good. One possible usability query I have is whether we should optimize for quickly flipping between several pages. Right now the UI is not optimized for this because of the scroll-to-top behavior when switching pages and because the Prev/Next actions are not next to each other. To clarify, I'm not suggesting a change here, just raising something that I think we need to look for user feedback on.

In general, I agree with this. I don't necessarily think the next/back placement is problematic at the bottom. But there could be better ways to move through pages more quickly, and controls at the top may be the right place for that.

> The hit area of each page number button could be increased. There is currently dead space between each one.

I can totally play around with that.

> The "Page $X of $Y threads" text in the top right looks wrong to me, because $X and $Y are different units. Showing "Threads $START to $END of $TOTAL" might work better? We could perhaps remove the total annotation count to make room for that. See Gmail's mail counter for a reference.

I'll explore this a bit. I am resistant to the `$START to $END of $TOTAL` because it further emphasizes thread-iness, could confuse users more (the number of results shown per page could be irregular, as the paging is based on threads), it requires pagination calculation in a place where it wouldn't yet have been calculated, and, most relevantly, it simply isn't "the done thing" in any examples I found, at all. No one's doing this anymore! And I think for good reason. It's chattery and distracting.

HOWEVER. Your apples-and-oranges unit-mismatch thing is real. I wonder if literally just getting rid of the word "of" could help here. I'd also like to split up the page number from the result count entirely once we move the user filter to the left...real estate issues. I'll experiment.

> When navigating through a large number of results, I definitely think it would be useful to have controls at the top, for the use case where someone is trying to skip to annotations made in some rough time period for example. These don't necessarily need to be the full controls, just Next/Prev arrows would be useful.

I can play a little more with this. FWIW, I had a hard time finding real-world examples of other places doing this, so please do send any examples you happen across my way.

**Comment by lyzadanger** [comment]
After discussion with the team: we've decided to limit the scope of this first chunk of pagination work to the controls at the bottom of the view. The challenges involved in the wording and pagination controls at the top of the view warrant their own focus (subsequent PR).

**Comment by lyzadanger** [comment]
Closing in lieu of #2994

**Pull request closed without merging**